consensus/system_contract/consensus.go

```diff
@@ -0,0 +1,390 @@
+	if number == 0 {
+		return nil
+	}
+	// Ensure that the block's timestamp isn't too close to its parent
+	var parent *types.Header
+	if len(parents) > 0 {
+		parent = parents[len(parents)-1]
+	} else {
+		parent = chain.GetHeader(header.ParentHash, number-1)
+	}
+	if parent == nil || parent.Number.Uint64() != number-1 || parent.Hash() != header.ParentHash {
+		return consensus.ErrUnknownAncestor
+	}
+	if header.Time < parent.Time {
+		return errInvalidTimestamp
+	}
+	// Verify that the gasUsed is <= gasLimit
+	if header.GasUsed > header.GasLimit {
+		return fmt.Errorf("invalid gasUsed: have %d, gasLimit %d", header.GasUsed, header.GasLimit)
+	}
+	if !chain.Config().IsCurie(header.Number) {
+		// Verify BaseFee not present before EIP-1559 fork.
+		if header.BaseFee != nil {
+			return fmt.Errorf("invalid baseFee before fork: have %d, want <nil>", header.BaseFee)
+		}
+		if err := misc.VerifyGaslimit(parent.GasLimit, header.GasLimit); err != nil {
+			return err
+		}
+	} else if err := misc.VerifyEip1559Header(chain.Config(), parent, header); err != nil {
+		// Verify the header's EIP-1559 attributes.
+		return err
+	}
+
+	// only if block header comes from a new block msg, then verify the signature against the current signer
+	if header.IsNewBlock {
+		signer, err := ecrecover(header)
+		if err != nil {
+			return err
+		}
+
+		s.lock.RLock()
+		defer s.lock.RUnlock()
+
+		if signer != s.signerAddressL1 {
+			log.Error("Unauthorized signer", "Got", signer, "Expected:", s.signerAddressL1)
+			return ErrUnauthorizedSigner
+		}
+	}
+
+	return nil
+}
+
+// VerifyUncles implements consensus.Engine, always returning an error for any
+// uncles as this consensus mechanism doesn't permit uncles.
+func (s *SystemContract) VerifyUncles(chain consensus.ChainReader, block *types.Block) error {
+	if len(block.Uncles()) > 0 {
+		return errors.New("uncles not allowed")
+	}
+	return nil
+}
+
+// Prepare initializes the consensus fields of a block header according to the
+// rules of a particular engine. Update only timestamp and prepare ExtraData for Signature
+func (s *SystemContract) Prepare(chain consensus.ChainHeaderReader, header *types.Header) error {
+	header.BlockSignature = make([]byte, extraSeal)
+	header.IsEuclidV2 = true
+	header.Extra = []byte{}
+	header.IsNewBlock = true
+	// Ensure the timestamp has the correct delay
+	parent := chain.GetHeader(header.ParentHash, header.Number.Uint64()-1)
+	if parent == nil {
+		return consensus.ErrUnknownAncestor
+	}
+	header.Time = parent.Time + s.config.Period
+	// If RelaxedPeriod is enabled, always set the header timestamp to now (ie the time we start building it) as
+	// we don't know when it will be sealed
+	if s.config.RelaxedPeriod || header.Time < uint64(time.Now().Unix()) {
+		header.Time = uint64(time.Now().Unix())
+	}
+	header.Difficulty = big.NewInt(1)
+	return nil
+}
+
+// Finalize implements consensus.Engine. There is no post-transaction
+// consensus rules in clique, therefore no rules here
```

Comment:
remove mentions of Clique here and everywhere in this file

Reply:
done.